dbt/include/sqlserver/macros/materializations/models/table/create_table_as.sql

```diff
@@ -1,26 +1,35 @@
 {% macro sqlserver__create_table_as(temporary, relation, sql) -%}
+   {#- TODO: add contracts here when in dbt 1.5 -#}
+   {%- set sql_header = config.get('sql_header', none) -%}
    {%- set as_columnstore = config.get('as_columnstore', default=true) -%}
-   {% set tmp_relation = relation.incorporate(
-   path={"identifier": relation.identifier.replace("#", "") ~ '_temp_view'},
-   type='view')-%}
    {%- set temp_view_sql = sql.replace("'", "''") -%}
+   {%- set tmp_relation = relation.incorporate(
+        path={"identifier": relation.identifier.replace("#", "") ~ '_temp_view'},
+        type='view') -%}
 
-   {{ sqlserver__drop_relation_script(tmp_relation) }}
+   {{- sql_header if sql_header is not none -}}
 
-   {{ sqlserver__drop_relation_script(relation) }}
+    -- drop previous temp view
+   {{- sqlserver__drop_relation_script(tmp_relation) }}
 
+    -- create temp view
    USE [{{ relation.database }}];
    EXEC('create view {{ tmp_relation.include(database=False) }} as
     {{ temp_view_sql }}
     ');
 
-   SELECT * INTO {{ relation }} FROM
-    {{ tmp_relation }}
+   -- select into the table and create it that way
+   {# TempDB schema is ignored, always goes to dbo #}
+   SELECT *
+   INTO {{ relation.include(database=False, schema=(not temporary))  }}
+   FROM {{ tmp_relation }}
```

Comment:
the mechanics of this are also interesting to me. is the order of operations around this
1. create `tmp_relation` which is a view
2. (this step) materialized this data into `dbo` schema
3. rename the resulting table into the correct database and schema?

Reply:
The operations here are there because SQL Server doesn't have CTAS (which Synapse does, so to be overridden there again). The closest thing in SQL Server is SELECT ... INTO target FROM source. But the source in the FROM clause there, cannot be a subquery. So this first creates a temporary view to serve as the FROM source as the view is the only thing which can just take the SQL as-is. The temporary view is dropped again at the end.

Reply:
The database=False is because we use USE [database] right above. The schema=(not temporary) is because tempdb only has dbo and will still use dbo if you put something else. So it would be confusing if you'd dive into the logs and then find out that the schema you've used doesn't exist (guess who had to do this).